### ISSUE TITLE
PerlIO::encoding::fallback corrupts filehandle

### ISSUE THREAD

**Issue by maros** [Perl/perl5#18495]
Module: PerlIO::encoding

**Description**
Setting the $PerlIO::encoding::fallback and using PerlIO::ecoding results in a broken filehandle that only returns the first 1024 characters of the file over and over. see https://perldoc.perl.org/PerlIO::encoding#DESCRIPTION

**Steps to Reproduce**
```
use PerlIO::encoding;
use Encode qw(:fallbacks);
$PerlIO::encoding::fallback = FB_PERLQQ; # or anything else like a CODEREF
open my $fh, "<:encoding(utf-8)","some_file.txt"; # some file with > 1024 characters
while (<$fh>) { print } 
```

**Expected behavior**
Print the content of the file, with all octets NOT mapping to utf-8 being replaced with \xHH (with HH being the hex representation of the octet). Instead the first 1024 characters of the file are printed over and over.

**Perl configuration**

```
Summary of my perl5 (revision 5 version 30 subversion 1) configuration:

  Platform:
    osname=linux
    osvers=4.19.0-11-amd64
    archname=x86_64-linux
    uname='linux gh-chroot-2020-11-04-1604493449-kqu4gwl 4.19.0-11-amd64 #1 smp debian 4.19.146-1 (2020-09-17) x86_64 gnulinux '
    config_args='-Dprefix=/opt/gh/perl-5.30.1-live -des -Dusedevel -Duseshrplib -Accflags=-DNO_TAINT_SUPPORT -Accflags=-mtune=core2 -Accflags=-fPIC -Accflags=-DPERL_DISABLE_PMC -Accflags=-ggdb -Dman1dir=none -Dman3dir=none -A'eval:scriptdir=/opt/gh/perl-5.30.1-live/bin''
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='cc'
    ccflags ='-DNO_TAINT_SUPPORT -mtune=core2 -fPIC -DPERL_DISABLE_PMC -ggdb -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2'
    cppflags='-DNO_TAINT_SUPPORT -mtune=core2 -fPIC -DPERL_DISABLE_PMC -ggdb -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='8.3.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/8/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.28.so
    so=so
    useshrplib=true
    libperl=libperl.so
    gnulibc_version='2.28'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E -Wl,-rpath,/opt/gh/perl-5.30.1-live/lib/5.30.1/x86_64-linux/CORE'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl):
  Compile-time options:
    HAS_TIMES
    NO_TAINT_SUPPORT
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DISABLE_PMC
    PERL_DONT_CREATE_GVSV
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
  Locally applied patches:
    Devel::PatchPerl 2.00
  Built under linux
  Compiled at Nov  4 2020 13:41:37
  %ENV:
    PERL5LIB="/opt/gh/geizhals-webapp/scripts/:/opt/gh/geizhals-importer/lib:/opt/gh/geizhals-webapp/gh-lib:/opt/gh/perl-current-live/local-default/lib/perl5:/opt/gh/geizhals-search/lib:/opt/gh/geizhals-importer/lib:/opt/gh/geizhals-webapp/gh-lib:/opt/gh/geizhals-kalif/lib:/opt/gh/geizhals-importer/lib:/opt/gh/geizhals-webapp/gh-lib:/opt/gh/geizhals-webapp:/opt/gh/geizhals-webapp/scripts/:/opt/gh/geizhals-importer/lib:/opt/gh/geizhals-webapp/gh-lib:/opt/gh/perl-current-live/local-default/lib/perl5:/opt/gh/geizhals-search/lib:/opt/gh/geizhals-importer/lib:/opt/gh/geizhals-webapp/gh-lib:/opt/gh/geizhals-kalif/lib:/opt/gh/geizhals-importer/lib:/opt/gh/geizhals-webapp/gh-lib:/opt/gh/geizhals-webapp:"
    PERL_CARTON_PATH="/opt/gh/perl-current-live/local-default"
  @INC:
    /opt/gh/geizhals-webapp/scripts/
    /opt/gh/geizhals-importer/lib
    /opt/gh/geizhals-webapp/gh-lib
    /opt/gh/perl-current-live/local-default/lib/perl5/5.30.1/x86_64-linux
    /opt/gh/perl-current-live/local-default/lib/perl5/5.30.1
    /opt/gh/perl-current-live/local-default/lib/perl5/x86_64-linux
    /opt/gh/perl-current-live/local-default/lib/perl5
    /opt/gh/geizhals-search/lib
    /opt/gh/geizhals-importer/lib
    /opt/gh/geizhals-webapp/gh-lib
    /opt/gh/geizhals-kalif/lib
    /opt/gh/geizhals-importer/lib
    /opt/gh/geizhals-webapp/gh-lib
    /opt/gh/geizhals-webapp
    /opt/gh/geizhals-webapp/scripts/
    /opt/gh/geizhals-importer/lib
    /opt/gh/geizhals-webapp/gh-lib
    /opt/gh/perl-current-live/local-default/lib/perl5/5.30.1/x86_64-linux
    /opt/gh/perl-current-live/local-default/lib/perl5/5.30.1
    /opt/gh/perl-current-live/local-default/lib/perl5/x86_64-linux
    /opt/gh/perl-current-live/local-default/lib/perl5
    /opt/gh/geizhals-search/lib
    /opt/gh/geizhals-importer/lib
    /opt/gh/geizhals-webapp/gh-lib
    /opt/gh/geizhals-kalif/lib
    /opt/gh/geizhals-importer/lib
    /opt/gh/geizhals-webapp/gh-lib
    /opt/gh/geizhals-webapp
    /opt/gh/perl-5.30.1-live/lib/site_perl/5.30.1/x86_64-linux
    /opt/gh/perl-5.30.1-live/lib/site_perl/5.30.1
    /opt/gh/perl-5.30.1-live/lib/5.30.1/x86_64-linux
    /opt/gh/perl-5.30.1-live/lib/5.30.1
```

**Comment by Leont** [comment]
`$PerlIO::encoding::fallback` MUST contain `Encode::STOP_AT_PARTIAL`, otherwise it will not work. I have a [branch](https://github.com/Perl/perl5/tree/smoke-me/leont/less-fragile-encoding) that enforces this, but apparently I haven't merged it yet.

**Comment by maros** [comment]
i retried with this option
```
$PerlIO::encoding::fallback = FB_PERLQQ | Encode::STOP_AT_PARTIAL;
```

as well as
```
$PerlIO::encoding::fallback = sub { return '?' };
```

with this file
```
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Tincidunt lobortis feugiat vivamus at augue eget arcu dictum varius. Nulla facilisi etiam dignissim diam quis enim. Sociis natoque penatibus et magnis dis parturient montes. Volutpat ac tincidunt vitae semper quis lectus nulla at. Feugiat in fermentum posuere urna nec tincidunt praesent semper. Gravida neque convallis a cras semper auctor. Vulputate sapien nec sagittis aliquam malesuada bibendum arcu. Arcu bibendum at varius vel pharetra vel turpis nunc eget. Sit amet luctus venenatis lectus magna.
Venenatis urna cursus eget nunc scelerisque viverra mauris in. Vitae congue eu consequat ac felis donec. Vitae purus faucibus ornare suspendisse. Feugiat in fermentum posuere urna nec tincidunt praesent. Ullamcorper a lacus vestibulum sed arcu non. Id velit ut tortor pretium. Aenean euismod elementum nisi quis. Purus in massa tempor nec feugiat nisl. A cras semper auctor neque vitae tempus. Arcu dui vivamus arcu felis bibendum. Suspendisse interdum consectetur libero id faucibus nisl tincidunt eget nullam. Adipiscing elit duis tristique sollicitudin nibh. Eros in cursus turpis massa tincidunt. A iaculis at erat pellentesque adipiscing commodo. Lectus magna fringilla urna porttitor rhoncus dolor purus. Sit amet nulla facilisi morbi tempus iaculis urna id.
Sit amet facilisis magna etiam tempor orci eu lobortis. Sed sed risus pretium quam vulputate dignissim suspendisse. Enim facilisis gravida neque convallis. Sed velit dignissim sodales ut eu sem integer vitae justo. Purus ut faucibus pulvinar elementum integer enim neque. Sed augue lacus viverra vitae congue eu consequat ac felis. Etiam sit amet nisl purus in mollis nunc sed id. Purus viverra accumsan in nisl. Sit amet luctus venenatis lectus. Vivamus at augue eget arcu. Cursus eget nunc scelerisque viverra mauris in. Lacus laoreet non curabitur gravida arcu ac tortor dignissim.
```

and just got
```
Venenatis urna cursus eget nunc scelerisque viverra mauris in. Vitae congue eu consequat ac felis donec. Vitae purus faucibus ornare suspendisse. Feugiat in fermentum posuere urna nec tincidunt praesent. Ull ....
```
over and over

**Comment by Leont** [comment]
It turns out that `FB_PERLQQ` also sets `LEAVE_SRC`, which is very much wrong in the case of `:encoding` (if it doesn't consume the input, it will keep repeating reading it). I guess you want to be using `PERLQQ` instead. IMO `:encoding` should force-ignore `LEAVE_SRC`, because that can't work anyway.

> as well as

Using a coderef as fallback is known to be problematic, because that doesn't allow for the passing of `STOP_AT_PARTIAL`, which means it will break if a multi-byte character ever bridges a buffer boundary.

**Comment by Leont** [comment]
This was fixed by `#18496`
